### DISCUSSION TITLE
Is it possible to suspend a running Rhai script and resume it later?

### DISCUSSION THREAD

**Discussion by casey** [rhaiscript/rhai#570] [Q&A] [answered]
We're thinking of using Rhai as the scripting language for a generative art toy thing. Users would write Rhai scripts that interact with the rendering engine to make cool images. One tricky requirement that we have is that we need a way to suspend user scripts after they draw a frame, return control to the browser so it can display the rendered frame, and then resume execution of the script where it left off. Something like this would be ideal:

```
// example user script

loop {
    render_frame();
    yield; // using an imaginary `yield` keyword that returns control to the rust caller, which can then resume execution later
}
```

Is something like this possible?

**Comment by schungx** [comment, marked as answer]
Not at this time.  You're asking for `async` (which is essentially suspend-and-resume).  Rhai is not async.

You have a few choices:

1) Use another scripting language that is async, such as https://github.com/rune-rs

2) Run the script in one thread, the rendering engine in another, using a channel to bind the two (check out the `threading` example)

3) Move your render loop outside of the script.  Call the script during each frame (or spawn a new engine during each frame)

**Reply by casey** [reply]
Thank you for the answer! Could we abuse debugger breakpoints to simulate async? I.e. create a function `yield_control()` that doesn't do anything, create a breakpoint on `yield_control`, suspend when we hit the breakpoint, and then resume later?

I think 2. isn't an option because we're in the browser, and there are no threads. Although maybe we can do that by running the Rhai user program in a web worker.

3. is hard because user scripts might have state, and we want to preserve that state between frames.

**Reply by schungx** [reply]
State is easy.  You just keep those states inside a `Scope` and pass that to the next call.  If user-kept state is what you're pondering over, it is a solved problem.

You may abuse the debugger, as a matter of fact.  Check out the `rhai-dbg` example.  It essentially do a single call and then do everything else within the debugger.  The only thing you need to make this happen is to set breakpoints on specific functions (like `render`) and then intercept it - but obviously, arguments will be a problem because you cannot get them without running the statement.

I am not sure that will work out for you.  Why can't you just render within the script?  Register a `render` function which draws to the browser and Rhai can call it to draw.  Why do you need to break out of Rhai's execution?

**Reply by casey** [reply]
If we don't return control to the browser, the page will become unresponsive, and newly rendered frames won't be displayed.

The steps we have to perform for each frame are:
- Call user-provided script to update rendering state
- Render
- Return control to the browser to render, process events and input, etc

We could have the user provide a script which renders each frame, and we just call it every frame, but then they wouldn't be able to persist their own state from frame to frame, and the script would start over from the beginning each time.

**Reply by schungx** [reply]
> We could have the user provide a script which renders each frame, and we just call it every frame, but then they wouldn't be able to persist their own state from frame to frame, and the script would start over from the beginning each time.

Why?  All the state will be kept in the `Scope`.  You can expose that as a mutable static variable to JavaScript, aka:

https://stackoverflow.com/questions/47722214/how-do-i-keep-internal-state-in-a-webassembly-module-written-in-rust

Never tried it myself, but your problem is how to maintain a persistent state in a WASM module.  If you Google it, I'm quite sure there is already a solution somewhere.

However, if you want to keep a dynamic property bag of states, that might be a problem as the WASM static is fix-sized.

**Reply by schungx** [reply]
Alternatively, if your states are small, you can just serialize to JSON and pass it back.  Then your front-end can pass that back to the next call and the `Scope` can be recreated from JSON.  That works only for very small states as you'll have to do the marshalling for every frame.

**Reply by casey** [reply]
Our problem isn't how to maintain persistent state in the WASM module, it's how to maintain the state of the user's script between frames. For example, if the user writes a Rhai script like this:

```
let x = expensive_init_step();

render_current_frame_with(x);
```

Then when we call it after every frame, it will re-initialize x, which is potentially expensive. Another issue is that the user might like some state that changes from frame to frame, and re-initializing it will reset it. For example:


```
let x = some_state()

render_current_frame_with(x);
x = next_state();
```

If we re-run the script from the beginning every frame, the state will never change.

**Reply by schungx** [reply]
Have you checked out these:

https://rhai.rs/book/engine/scope.html

https://rhai.rs/book/patterns/events.html

These are examples on how to keep a bunch of states across different calls to the engine.

So, in other words,  your problem IS to maintain a _persistent_ state across calls (i.e. frames).

**Reply by schungx** [reply]
In fact, the scriptable event handler pattern is exactly what you want.

Think about it, you _ARE_ writing an event handler in WASM.  The event is `render_frame` which obviously gets called 60 times a second.

**Reply by casey** [reply]
That looks great, I didn't see those. Is there an example of the user scripts that go along with the Rust code? I'm assuming the user script would just define `start`, `end`, and `update` functions, like this:


```
let x = some_initial_state();

fn start() {
}

fn end() {
}

fn update() {
  // can use and mutate `x`
}
```

**Reply by schungx** [reply]
There are three actual implementations in the `examples` folder, one for each scripting style.

**Reply by casey** [reply]
Awesome, thank you, I'll check those out!

**Reply by schungx** [reply]
> That looks great, I didn't see those. Is there an example of the user scripts that go along with the Rust code? I'm assuming the user script would just define `start`, `end`, and `update` functions, like this:

Usually people like to keep it consistent by having an `init` function that is called once at the very beginning, and then maybe a `finish` or `destroy` function that is called at the very end.

However, if you want to make the global portion of the script a de-facto `init` function you can also do that, just make sure to disable running the script before each call (otherwise your `init` function is essentially run each time).

**Comment by schungx** [comment]
Closing this for now. Feel free to reopen if necessary.
